File: docs/src/python/check_constraints.py
def add_constraint():
    # --8<-- [start:add_constraint]
    from deltalake import DeltaTable

    dt = DeltaTable("../rust/tests/data/simple_table")

    # Check the schema before hand
    print(dt.schema())
    # Add the constraint to the table.
    dt.alter.add_constraint({"id_gt_0": "id > 0"})
    # --8<-- [end:add_constraint]


def add_data():
    # --8<-- [start:add_data]
    from deltalake import write_deltalake, DeltaTable
    import pandas as pd

    dt = DeltaTable("../rust/tests/data/simple_table")

    df = pd.DataFrame({"id": [-1]})
    write_deltalake(dt, df, mode="append", engine="rust")
    # _internal.DeltaProtocolError: Invariant violations: ["Check or Invariant (id > 0) violated by value in row: [-1]"]
    # --8<-- [end:add_data]

File: docs/src/rust/check_constraints.rs
use std::{path::Path, sync::Arc};
use url::Url;

use arrow_array::{Array, Int32Array, RecordBatch};
use deltalake::{writer::{DeltaWriter, RecordBatchWriter}};


#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {

    // --8<-- [start:add_constraint]
    let path = Path::new("../rust/tests/data/simple_table").canonicalize().unwrap();
    let delta_path = Url::from_directory_path(path).unwrap();
    let table = deltalake::open_table(delta_path).await?;
    
    table.add_constraint().with_constraint("id_gt_0", "id > 0").await?;
    // --8<-- [end:add_constraint]

    // --8<-- [start:add_data]
    let mut table = deltalake::open_table(delta_path).await?;
    let schema = table.snapshot()?.snapshot().arrow_schema();
    let invalid_values: Vec<Arc<dyn Array>> = vec![
        Arc::new(Int32Array::from(vec![-10]))
    ];
    let batch = RecordBatch::try_new(schema, invalid_values)?;
    let mut writer = RecordBatchWriter::for_table(&table)?;
    writer.write(batch).await?;
    writer.flush_and_commit(&mut table).await?;
    // --8<-- [end:add_data]

    Ok(())
}